### PR TITLE
Web Backend Recover Wallet

### PR DESCRIPTION
**Pull request by ahadda5**
**What type of PR is this?**
 Feature  [#8664](https://github.com/prysmaticlabs/prysm/issues/8664)

**What does this PR do? Why is it needed?**
Web UI is being updated to be an alternative to recover wallet CLI.  In doing so, the backend has to handle the recover Wallet request. 

**Which issues(s) does this PR fix?**

Part of [#8664 ](https://github.com/prysmaticlabs/prysm/issues/8664) The portion about using the supported bip39 Mnemonic language in recovering their wallet from a mnemonic. 

**Other notes for review**
While that portion of the issue started as adding the Menomic language support the following modifications were done 

1. Recovering from wallet involves using an optional 25th word.  So the request has to accommodate for whether the user wants that or not  ( a flag to skip or not and the paraphrase itself).  The accounts.RecoverWalletConfig  struct used in recovering, either specifies the paraphrase if not to be skipped or keeps it nil. 
2. The rpc CreateWallet was divided into the three types of Wallets  (Imported , Derived or Remote).  The imported stayed part of the rpc.CreateWallet, while the derived is the rpc.RecoverWallet with the added checks and balances(see below) to handle the passed parameters.   
In short, Either you import keystores, or you recover a wallet. (remote are not supported through the web) 
3. Test cases cover :   passing a supported language from the web,  passing a 25th mnemonic(if not to be skipped) , validating the passed password ( the check between a password and its confirm is part of the web frontend), validating the numAccounts(atleast 1) . Once that passes, the RecoverWalletConfig  struct is passed to the legacy accounts.RecoverWallet

### CODE REVIEW

File: proto/beacon/db/finalized_block_root_container.pb.go [hunk 1]
```diff
@@ -5,11 +5,10 @@ package db
 
 import (
 	fmt "fmt"
+	proto "github.com/gogo/protobuf/proto"
```
Comment (terencechain) [inline proto/beacon/db/finalized_block_root_container.pb.go:8]: Can you run goimport?

File: proto/validator/accounts/v2/web_api.proto [hunk 2]
```diff
@@ -200,6 +206,15 @@ message WalletResponse {
     KeymanagerKind keymanager_kind = 2;
 }
 
+message RecoverWalletRequest {
+    string mnemonic = 1;
+    uint64 num_accounts = 2;
+    string wallet_password = 3;
+    string language = 4;
+    string mnemonic25thWord =5 ;
+    bool   skipMnemonic25thWord = 6;
```
Comment (terencechain) [inline proto/validator/accounts/v2/web_api.proto:215]:
```suggestion
    bool   skip_mnemonic_25th_word = 6; 
```

File: proto/validator/accounts/v2/web_api.proto [hunk 3]
```diff
@@ -200,6 +206,15 @@ message WalletResponse {
     KeymanagerKind keymanager_kind = 2;
 }
 
+message RecoverWalletRequest {
+    string mnemonic = 1;
+    uint64 num_accounts = 2;
+    string wallet_password = 3;
+    string language = 4;
+    string mnemonic25thWord =5 ;
```
Comment (terencechain) [inline proto/validator/accounts/v2/web_api.proto:214]:
```suggestion
    string mnemonic_25th_word =5 ;
```

File: validator/rpc/wallet.go [hunk 4]
```diff
@@ -165,6 +133,80 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
 	}, nil
 }
 
+// RecoverWallet via an API request, allowing a user to recover a derived.
+// Generate the seed from the mnemonic + language
+// Create N validator keystores from the seed specified by req.NumAccounts
+// Set the wallet password to req.WalletPassword (needs password validation)
+// Create the wallet
+// Return &CreateWalletResponse, nil if nothing went wrong
+func (s *Server) RecoverWallet(ctx context.Context, req *pb.RecoverWalletRequest) (*pb.CreateWalletResponse, error) {
```
Comment (terencechain) [inline validator/rpc/wallet.go:142]: This function needs test

File: validator/rpc/wallet.go [hunk 5]
```diff
@@ -165,6 +133,80 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
 	}, nil
 }
 
+// RecoverWallet via an API request, allowing a user to recover a derived.
```
Comment (terencechain) [inline validator/rpc/wallet.go:136]: Can we rephrase the function comment for readability?

**Review by terencechain — CHANGES_REQUESTED** [review]

File: proto/beacon/db/finalized_block_root_container.pb.go (hunk 1 above)
Comment (ahadda5) [reply proto/beacon/db/finalized_block_root_container.pb.go:8]: Ahh!   good to know.  Cause running` ./scripts/update-go-pbs.sh ` didn't fail   but i just noticed it ended with 
./scripts/update-go-pbs.sh: line 28: goimports: command not found
I'm investigating .

Comment (ahadda5) [reply proto/beacon/db/finalized_block_root_container.pb.go:8]: goimports installed (golang x tools)

File: validator/rpc/wallet.go (hunk 4 above)
Comment (ahadda5) [reply validator/rpc/wallet.go:141]: This functions was testing using an adaptation of the func [TestServer_CreateWallet_Derived](https://github.com/prysmaticlabs/prysm/blob/4886a4e749b16adb1725004cee39661c6e10f00e/validator/rpc/wallet_test.go#L86)
The Test func was adapted to handle passing a language, mnemonic, skip25pass flag and the passphrase.

Comment (ahadda5) [reply validator/rpc/wallet.go:141]: Did i miss other scenarios ?

File: validator/rpc/wallet_test.go [hunk 6]
```diff
@@ -83,34 +83,55 @@ func TestServer_CreateWallet_Imported(t *testing.T) {
...
+	_, err = s.RecoverWallet(ctx, req)
+	require.ErrorContains(t, "invalid mnemonic in request", err)
 
 	mnemonicResp, err := s.GenerateMnemonic(ctx, &empty.Empty{})
 	require.NoError(t, err)
 	req.Mnemonic = mnemonicResp.Mnemonic
 
-	_, err = s.CreateWallet(ctx, req)
+	req.Mnemonic25ThWord = " "
+	req.SkipMnemonic_25ThWord = false
```
Comment (terencechain) [inline validator/rpc/wallet_test.go:117]:
```suggestion
	req.SkipMnemonic25thWord = false
```

File: validator/rpc/wallet_test.go [hunk 7]
```diff
@@ -83,34 +83,55 @@ func TestServer_CreateWallet_Imported(t *testing.T) {
...
+	req.Language = "english"
+	_, err = s.RecoverWallet(ctx, req)
+	require.ErrorContains(t, "invalid mnemonic in request", err)
 
 	mnemonicResp, err := s.GenerateMnemonic(ctx, &empty.Empty{})
 	require.NoError(t, err)
 	req.Mnemonic = mnemonicResp.Mnemonic
 
-	_, err = s.CreateWallet(ctx, req)
+	req.Mnemonic25ThWord = " "
```
Comment (terencechain) [inline validator/rpc/wallet_test.go:116]:
```suggestion
	req.Mnemonic25thWord = " "
```

File: validator/rpc/wallet.go [hunk 8]
```diff
@@ -165,6 +133,79 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+	//accounts.Recoverwallet checks if wallet already exists
+	walletDir := s.walletDir
+
+	//web-ui should check the new and confirmed password are equal
+	walletPassword := req.WalletPassword
+	if err := promptutil.ValidatePasswordInput(walletPassword); err != nil {
+		return nil, status.Error(codes.InvalidArgument, "password did not pass validation")
+	}
+
+	//recover
```
Comment (terencechain) [inline validator/rpc/wallet.go:181]:
```suggestion
```

File: validator/rpc/wallet.go [hunk 9]
```diff
@@ -165,6 +133,79 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+	}
+	if !req.SkipMnemonic_25ThWord && strings.TrimSpace(req.Mnemonic25ThWord) == "" {
+		return nil, status.Error(codes.InvalidArgument, "mnemonic 25th word passphrase cannot be empty")
+	}
+
+	//web UI is structured to only write to the default wallet directory
+	//accounts.Recoverwallet checks if wallet already exists
+	walletDir := s.walletDir
+
+	//web-ui should check the new and confirmed password are equal
```
Comment (terencechain) [inline validator/rpc/wallet.go:175]:
```suggestion
	// Web-ui should check the new and confirmed password are equal
```

File: validator/rpc/wallet.go [hunk 10]
```diff
@@ -165,6 +133,79 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+	bip39.SetWordList(allowedLanguages[language])
+	mnemonic := req.Mnemonic
+	if err := accounts.ValidateMnemonic(mnemonic); err != nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid mnemonic in request")
+	}
+	if !req.SkipMnemonic_25ThWord && strings.TrimSpace(req.Mnemonic25ThWord) == "" {
+		return nil, status.Error(codes.InvalidArgument, "mnemonic 25th word passphrase cannot be empty")
+	}
+
+	//web UI is structured to only write to the default wallet directory
```
Comment (terencechain) [inline validator/rpc/wallet.go:171]:
```suggestion
	// Web UI is structured to only write to the default wallet directory
```

File: validator/rpc/wallet.go [hunk 11]
```diff
@@ -165,6 +133,79 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+// Create N validator keystores from the seed specified by req.NumAccounts.
+// Set the wallet password to req.WalletPassword, then create the wallet from
+// the provided Mnemonic and return CreateWalletResponse,nil if nothing went wrong
+func (s *Server) RecoverWallet(ctx context.Context, req *pb.RecoverWalletRequest) (*pb.CreateWalletResponse, error) {
+	numAccounts := int(req.NumAccounts)
+	if numAccounts < 1 {
+		return nil, status.Error(codes.InvalidArgument, "Must create at least 1 validator account")
+	}
+
+	//check validate mnemonic with chosen language
```
Comment (terencechain) [inline validator/rpc/wallet.go:147]:
```suggestion
	// Check validate mnemonic with chosen language
```

File: validator/rpc/wallet.go [hunk 12]
```diff
@@ -165,6 +133,79 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+		"chinese_simplified":  wordlists.ChineseSimplified,
+		"chinese_traditional": wordlists.ChineseTraditional,
+		"french":              wordlists.French,
+		"italian":             wordlists.Italian,
+		"japanese":            wordlists.Japanese,
+		"korean":              wordlists.Korean,
+		"spanish":             wordlists.Spanish,
+	}
+	if _, ok := allowedLanguages[language]; !ok {
+		return nil, status.Error(codes.InvalidArgument, "input not in the list of allowed languages")
```
Comment (terencechain) [inline validator/rpc/wallet.go:160]:
```suggestion
		return nil, status.Error(codes.InvalidArgument, "input not in the list of supported languages")
```

File: validator/rpc/wallet.go [hunk 13]
```diff
@@ -165,6 +133,79 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+	}
+	bip39.SetWordList(allowedLanguages[language])
+	mnemonic := req.Mnemonic
+	if err := accounts.ValidateMnemonic(mnemonic); err != nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid mnemonic in request")
+	}
+	if !req.SkipMnemonic_25ThWord && strings.TrimSpace(req.Mnemonic25ThWord) == "" {
+		return nil, status.Error(codes.InvalidArgument, "mnemonic 25th word passphrase cannot be empty")
```
Comment (terencechain) [inline validator/rpc/wallet.go:168]:
```suggestion
		return nil, status.Error(codes.InvalidArgument, "mnemonic 25th word cannot be empty")
```

File: validator/rpc/wallet.go [hunk 14]
```diff
@@ -165,6 +133,79 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+	}
+
+	//web UI is structured to only write to the default wallet directory
+	//accounts.Recoverwallet checks if wallet already exists
```
Comment (terencechain) [inline validator/rpc/wallet.go:172]:
```suggestion
	// accounts.Recoverwallet checks if wallet already exists
```

File: validator/rpc/wallet.go (hunk 4 above)
Comment (terencechain) [reply validator/rpc/wallet.go:141]: The following scenarios are missed, it'll be great to cover them
1.) `password did not pass validation`
2.) `Could not write wallet password to disk`

File: validator/rpc/wallet_test.go (hunk 6 above)
Comment (ahadda5) [reply validator/rpc/wallet_test.go:117]: Ok.  So for SkipMnemonic25thWord  , it has to be defined in proto as skip_mnemonic_25th_word.   I'll do like we did with mnemonic_25th_word up .

File: proto/validator/accounts/v2/web_api.proto (hunk 3 above)
Comment (ahadda5) [reply proto/validator/accounts/v2/web_api.proto:214]: So    Mnemonic25ThWord is the result of mnemonic25th_word   and
Mnemonic_25ThWord is the result of  mnemonic_25th_word     
So after the numbers it upper cases the First letter  and after the underscore it does the same.  If faced by _ and a number it keeps the _.  
I'll maintain   mnemonic25th_word  and skip_mnemonic25th_word  .

File: validator/rpc/wallet.go (hunk 5 above)
Comment (ahadda5) [reply validator/rpc/wallet.go:136]: done

File: proto/validator/accounts/v2/web_api.proto [hunk 15]
```diff
@@ -200,6 +206,15 @@ message WalletResponse {
     KeymanagerKind keymanager_kind = 2;
 }
 
+message RecoverWalletRequest {
+    string mnemonic = 1;
+    uint64 num_accounts = 2;
+    string wallet_password = 3;
+    string language = 4;
+    string mnemonic25th_word =5 ;
```
Comment (nisdas) [inline proto/validator/accounts/v2/web_api.proto:214]:
```suggestion
    string mnemonic25th_word = 5 ;
```

File: validator/accounts/wallet_recover.go [hunk 16]
```diff
@@ -219,7 +219,7 @@ func inputNumAccounts(cliCtx *cli.Context) (int64, error) {
 	return int64(numAccountsInt), nil
 }
 
-func validateMnemonic(mnemonic string) error {
+func ValidateMnemonic(mnemonic string) error {
```
Comment (nisdas) [inline validator/accounts/wallet_recover.go:222]: please add in function comment here if you are exporting the function.

File: validator/rpc/wallet.go [hunk 17]
```diff
@@ -165,6 +133,78 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
 }
 
+// RecoverWallet via an API request, allowing a user to recover a derived.
+// Generate the seed from the mnemonic + language + 25th passphrase(optional).
+// Create N validator keystores from the seed specified by req.NumAccounts.
+// Set the wallet password to req.WalletPassword, then create the wallet from
+// the provided Mnemonic and return CreateWalletResponse,nil if nothing went wrong
+func (s *Server) RecoverWallet(ctx context.Context, req *pb.RecoverWalletRequest) (*pb.CreateWalletResponse, error) {
+	numAccounts := int(req.NumAccounts)
+	if numAccounts < 1 {
```
Comment (nisdas) [inline validator/rpc/wallet.go:143]:
```suggestion
	if numAccounts == 0 {
```
for a clearer validation condition.

File: validator/rpc/wallet.go [hunk 18]
```diff
@@ -165,6 +133,78 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+		"english":             wordlists.English,
+		"chinese_simplified":  wordlists.ChineseSimplified,
+		"chinese_traditional": wordlists.ChineseTraditional,
+		"french":              wordlists.French,
+		"italian":             wordlists.Italian,
+		"japanese":            wordlists.Japanese,
+		"korean":              wordlists.Korean,
+		"spanish":             wordlists.Spanish,
+	}
+	if _, ok := allowedLanguages[language]; !ok {
```
Comment (nisdas) [inline validator/rpc/wallet.go:159]: if the language variable is 
```
English
```
```
Korean
```
this request will fail, the validation should be capitalization agnostic.

File: proto/validator/accounts/v2/web_api.proto [hunk 19]
```diff
@@ -200,6 +206,15 @@ message WalletResponse {
...
+    bool   skip_mnemonic25th_word = 6;
```
Comment (nisdas) [inline proto/validator/accounts/v2/web_api.proto:215]: a bool here seems redundant. If the request has `mnemonic25th_word == '' ` , it is a reasonable assumption that the requester is skipping the 25th word.

File: validator/rpc/wallet.go [hunk 20]
```diff
@@ -165,6 +133,78 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
...
+		return nil, err
+	}
+	if err := s.initializeWallet(ctx, &wallet.Config{
+		WalletDir:      walletDir,
+		KeymanagerKind: keymanager.Derived,
+		WalletPassword: walletPassword,
+	}); err != nil {
+		return nil, err
+	}
+	if err := writeWalletPasswordToDisk(walletDir, walletPassword); err != nil {
```
Comment (nisdas) [inline validator/rpc/wallet.go:197]: why are writing the password to disk ? Isn't this something we are moving away from 
cc @rauljordan

**Review by nisdas — CHANGES_REQUESTED** [review]
Please use a better name for the PR title, its not immediately obvious what the PR is trying to accomplish as all the words are compressed into a single one.

File: validator/rpc/wallet.go (hunk 4 above)
Comment (ahadda5) [reply validator/rpc/wallet.go:141]: First password test added and passed, since it was new code. 
As for the write wallet password to disk. Glad you pointed it out.  The already existing Create Imported does the same . So adding,  at the end , the following lines(taken from the [Test_writeWalletPasswordtoDisk )](url) to both Recover (derived) and Create Imported  causes the test to fail.  Need to investigate. 


```
       // File should have been written.
	assert.Equal(t, true, fileutil.FileExists(localWalletDir))

	// Attempting to write again should trigger an error.
	err = writeWalletPasswordToDisk(localWalletDir, "somepassword")
	require.NotNil(t, err)
```

File: proto/validator/accounts/v2/web_api.proto (hunk 19 above)
Comment (ahadda5) [reply proto/validator/accounts/v2/web_api.proto:215]: So Web UI handles whether to skip or not and if not passes the phrase. Backend rpc only worries to check that it is not an empty string (strings.TrimSpace(mnemonic) != '') ?

Comment (nisdas) [reply proto/validator/accounts/v2/web_api.proto:215]: That is correct, thats how it looks like from my point of view. I am not privy to any previous discussions @rauljordan might have had with you on this regarding the schema but it seems like we can be more efficient here.

Comment (rauljordan) [reply proto/validator/accounts/v2/web_api.proto:215]: We haven't implemented the 25th word feature on the frontend, so I think it is ok for it to be optional. The backend just has to check it is an empty string, so I would get rid of this bool in the request

File: validator/rpc/wallet.go [hunk 21]
```diff
@@ -30,7 +33,7 @@ const (
 )
 
 // CreateWallet via an API request, allowing a user to save a new
-// derived, imported, or remote wallet.
+// imported (derived is RecoverWallet and remote is not supported through web)
```
Comment (rauljordan) [inline validator/rpc/wallet.go:36]:
```suggestion
// imported wallet via RPC
```

File: validator/rpc/wallet.go [hunk 22]
```diff
@@ -165,6 +133,78 @@ func (s *Server) WalletConfig(ctx context.Context, _ *empty.Empty) (*pb.WalletRe
 	}, nil
 }
 
+// RecoverWallet via an API request, allowing a user to recover a derived.
+// Generate the seed from the mnemonic + language + 25th passphrase(optional).
+// Create N validator keystores from the seed specified by req.NumAccounts.
+// Set the wallet password to req.WalletPassword, then create the wallet from
+// the provided Mnemonic and return CreateWalletResponse,nil if nothing went wrong
```
Comment (rauljordan) [inline validator/rpc/wallet.go:140]:
```suggestion
// the provided Mnemonic and return CreateWalletResponse.
```

File: validator/rpc/wallet.go (hunk 20 above)
Comment (rauljordan) [reply validator/rpc/wallet.go:197]: I believe we write the password to disk if a feature flag is enabled to help companies such as dappnode start their validator from the password file if needed. It does not get saved for regular users unless they add this flag. Since this conditional is hidden within the helper function, it looks opaque from the perspective of the reader, so perhaps we should do something like:

```go
if featureconfig.Get().WriteWalletPasswordToDisk {
  writeWalletPasswordToDisk
}
```
Up to you if you want to do that change in this PR @ahadda5

File: validator/rpc/wallet_test.go [hunk 23]
```diff
@@ -83,34 +83,62 @@ func TestServer_CreateWallet_Imported(t *testing.T) {
...
+	req.SkipMnemonic25ThWord = false
+	_, err = s.RecoverWallet(ctx, req)
+	require.ErrorContains(t, "mnemonic 25th word cannot be empty", err)
+	req.Mnemonic25ThWord = "outer"
+
+	//test weak password
```
Comment (rauljordan) [inline validator/rpc/wallet_test.go:123]:
```suggestion
	// Test weak password.
```

File: validator/rpc/wallet_test.go [hunk 24]
```diff
@@ -83,34 +83,62 @@ func TestServer_CreateWallet_Imported(t *testing.T) {
...
+	req.WalletPassword = "123qwe"
+	_, err = s.RecoverWallet(ctx, req)
+	require.ErrorContains(t, "password did not pass validation", err)
+
+	req.WalletPassword = strongPass
+	//create(drived should fail) then test recover
```
Comment (rauljordan) [inline validator/rpc/wallet_test.go:129]: Let's try to use consistent formatting for comments, with capitalization and punctuation with spacing as needed

**Review by Oleksandr-prog — APPROVED** [review]

File: validator/rpc/wallet.go (hunk 20 above)
Comment (ahadda5) [reply validator/rpc/wallet.go:197]: But that is already part of the function itself. 
if  !writeWalletPassowrdonWebOnBoarding  don't write.   Still need it?

Comment (ahadda5) [reply validator/rpc/wallet.go:197]: @rauljordan  check above [Discussion](https://github.com/prysmaticlabs/prysm/pull/8679#discussion_r603139561)

File: validator/rpc/wallet.go (hunk 4 above)
Comment (ahadda5) [reply validator/rpc/wallet.go:141]: So after @rauljordan  pointed out the flag i added it to the test case. But Recover is still not writing. 
I update the flag to True and expect RecoverWallet to write the password,  but it does not. 
Then a defer call to set to False.

File: validator/rpc/wallet_test.go [hunk 25]
```diff
@@ -134,11 +133,30 @@ func TestServer_RecoverWallet_Derived(t *testing.T) {
...
+	resetCfgFalse := featureconfig.InitWithReset(&featureconfig.Flags{
+		WriteWalletPasswordOnWebOnboarding: false,
+	})
+	defer resetCfgFalse()
 
+	resetCfgTrue := featureconfig.InitWithReset(&featureconfig.Flags{
+		WriteWalletPasswordOnWebOnboarding: true,
+	})
+	resetCfgTrue()
+
```
Comment (ahadda5) [inline validator/rpc/wallet_test.go]: Need to figure out this test; basically how to set the writing and have Recover/Create write the password.

File: proto/validator/accounts/v2/web_api.pb.go [hunk 26]
```diff
@@ -18,6 +14,9 @@ import (
 	grpc "google.golang.org/grpc"
 	codes "google.golang.org/grpc/codes"
 	status "google.golang.org/grpc/status"
+	io "io"
```
Comment (terencechain) [inline proto/validator/accounts/v2/web_api.pb.go:17]: Please revert these lines. Go import will fix this

File: proto/validator/accounts/v2_gateway/web_api.pb.go [hunk 27]
```diff
@@ -21,6 +18,8 @@ import (
 	status "google.golang.org/grpc/status"
 	protoreflect "google.golang.org/protobuf/reflect/protoreflect"
 	protoimpl "google.golang.org/protobuf/runtime/protoimpl"
+	reflect "reflect"
```
Comment (terencechain) [inline proto/validator/accounts/v2_gateway/web_api.pb.go:21]: Same here

File: validator/rpc/wallet.go (hunk 4 above)
Comment (terencechain) [reply validator/rpc/wallet.go:141]: Awesome. We just need a test for the following error condition then we should be good to go:
```go
validatingPublicKeys, err := km.FetchValidatingPublicKeys(ctx)
	if err != nil {
		return errors.Wrap(err, "could not check for validating public keys")
	}
```

Comment (ahadda5) [reply validator/rpc/wallet.go:141]: @terencechain  this piece of code was not added (pre this PR) and is already existed  tested in [TestServer_ImportKeystores_OK](https://github.com/ahadda5/prysm/blob/2bdb65e668d4cc205303ef4b0099bc69f59395a8/validator/rpc/wallet_test.go#L257)   It initializes a wallet, a child keymanager and fills it with keystores. It fetches before importing the keystore, then fetches the public keys again.
  Is there a something specific you would like me to add?

File: proto/validator/accounts/v2/web_api.pb.go (hunk 26 above)
Comment (ahadda5) [reply proto/validator/accounts/v2/web_api.pb.go:17]: So the way i would do is, to open the file and move those 3 lines up to their original place?  it is a do not edit file.

Comment (terencechain) [reply proto/validator/accounts/v2/web_api.pb.go:17]: Or run `goimports` that will format these files for you

File: validator/rpc/wallet.go (hunk 4 above)
Comment (terencechain) [reply validator/rpc/wallet.go:141]: The following error condition is not tested according to code coverage though:
```
could not check for validating public keys
```

I think that something that we can add with relatively low effort

**Review by nisdas — APPROVED** [review]
Looks good to me but the PR has conflicts

File: validator/rpc/wallet_test.go [hunk 28]
```diff
@@ -83,35 +83,80 @@ func TestServer_CreateWallet_Imported(t *testing.T) {
...
+	_, err = s.RecoverWallet(ctx, req)
 	require.NoError(t, err)
+
+	// Password File should have been written.
+	passwordFilePath := filepath.Join(localWalletDir, wallet.DefaultWalletPasswordFile)
+	assert.Equal(t, true, fileutil.FileExists(passwordFilePath))
+
+	// Attempting to write again should trigger an error.
+	err = writeWalletPasswordToDisk(localWalletDir, "somepassword")
+	require.NotNil(t, err)
```
Comment (terencechain) [inline validator/rpc/wallet_test.go:158]:
```suggestion
	require.ErrorContains(t, "cannot write wallet password file as it already exists", err)
```

**Review by terencechain — APPROVED** [review]

**Comment by rauljordan** [comment]
Hi @ahadda5 heads-up, adding `Fixes #` will close the particular issue. Typically, we just say "Part of #issuenum` when the issue is still in progress. I edited your description to reflect this

**Comment by rauljordan** [comment]
Also you don't need to add a markdown hyperlink to the issue number. Github will automatically do it if you just type in #8129 for example
